### PR TITLE
Separates tile request from renderer logic

### LINKED ISSUE #117: Renderers shouldn't request tiles directly
Right now, on the server, the renderers are directly requesting and using tiles.

This causes a number of problems:
* It makes it much harder to cache tiles
* Courseness code needs to be worked into each renderer
* It just generally makes the renderers much more complex than they need to be.

The Tile Service should be requesting the tiles and passing them to the renderer.

### PR DESCRIPTION
**Pull request by robharper**
This PR takes the tile request logic out of the individual renderers and moves the responsibility to the caller, the TileService implementation.

The ramifications are that the renderers now advertise (via *Accepts methods) which data type in the TileData they accept and are generified as such. The concept of coarseness leads to the selection of a tile at a lower (more zoomed out) level that encompasses the requested tile wrapped in a view (TileDataView). The view makes the data from the parent tile look like the requested tile.

Overall this PR reduces code duplication and makes renderers slightly closer to single concern.

Fixes #117

### CODE REVIEW

File: tile-service/src/main/java/com/oculusinfo/tile/rendering/ImageRendererFactory.java [hunk 1]
```diff
@@ -90,11 +90,6 @@ protected TileDataImageRenderer create () {
 			return new TopAndBottomTextScoresImageRenderer();
 		} else if ("doublestatistics".equals(rendererType)) {
 			return new DoublesStatisticImageRenderer();
-		} else if ("metadata".equals(rendererType)) {
-		    List<String> components = getPropertyValue(METADATA_COMPONENTS);
```
Comment (nkronenfeld) [inline tile-service/src/main/java/com/oculusinfo/tile/rendering/ImageRendererFactory.java]: The three properties used by the MetaDataRenderer are no longer needed; they should be removed too.

File: tile-service/src/main/java/com/oculusinfo/tile/rest/tile/TileServiceImpl.java [hunk 2]
```diff
@@ -77,20 +124,33 @@ public BufferedImage getTileImage( String layer, TileIndex index, Iterable<TileI
...
             // produce the tile renderer from the configuration
 			TileDataImageRenderer tileRenderer = config.produce(TileDataImageRenderer.class);
             // prepare for rendering
 			config.prepareForRendering(layer, index, tileSet);
-            // render to buffered image
-			bi = tileRenderer.render(config);
+
+			String dataId = config.getPropertyValue(LayerConfiguration.DATA_ID);
+			PyramidIO pyramidIO = config.produce(PyramidIO.class);
+			TileSerializer<Double> serializer = SerializationTypeChecker.checkBinClass(config.produce(TileSerializer.class),
```
Comment (nkronenfeld) [inline tile-service/src/main/java/com/oculusinfo/tile/rest/tile/TileServiceImpl.java]: Should not be <Double>

**Comment by nkronenfeld** [comment]
I see a couple issues with TileDataView:
* It does not contain any logic pertaining to the width of the wrapping tile, neither in construction nor in getBin.
* setBin and getData also need to be overridden, even if it's just to throw an UnsupportedOperationException.

File: tile-service/src/main/java/com/oculusinfo/tile/rest/tile/TileServiceImpl.java (hunk 2 above)
Comment (robharper) [inline tile-service/src/main/java/com/oculusinfo/tile/rest/tile/TileServiceImpl.java]: Sorry, did this comment get cut off?

**Comment by robharper** [comment]
The width of the wrapping tile as in bin count? It is derived from the source tile and the level delta like so:
```
sourceIndex.getXBins()/tileCountRatio
```

getBin should check that you are within tile bounds, yes, because unlike a regular tile where you'd go out of bounds you'd actually get real data back even though you are out of bounds.

File: tile-service/src/main/java/com/oculusinfo/tile/rest/tile/TileServiceImpl.java (hunk 2 above)
Comment (nkronenfeld) [inline tile-service/src/main/java/com/oculusinfo/tile/rest/tile/TileServiceImpl.java]: No, but it treated the lt and gt signs as tag markers and didn't show them

**Comment by nkronenfeld** [comment]
Are you able to do this today? If not, I can just make the changes myself and pull

**Comment by robharper** [comment]
Working on it now. Adding test cases

File: tile-service/src/main/java/com/oculusinfo/tile/rendering/ImageRendererFactory.java (hunk 1 above)
Comment (robharper) [inline tile-service/src/main/java/com/oculusinfo/tile/rendering/ImageRendererFactory.java]: Fixed

**Comment by nkronenfeld** [comment]
error-checking added to getBin needs to be added to setBin too.

**Comment by robharper** [comment]
Done.

Arguably the bounds checking should live in TileData itself since there is currently no error checking. The checks in TileDataView would work copy-paste into TileData. At the moment the x*width + y index calculation could lead to strange behavior.

**Comment by nkronenfeld** [comment]
Good point - do you want to do that here, or separately?

**Comment by robharper** [comment]
I'll add it to this PR since its need was identified by the use case created here.

**Comment by robharper** [comment]
Unfortunately the checks can't be moved and have to be duplicated in the TileData and TileDataViews. The reason is the view wraps a source tile data with different bin counts so something in range in the source can be out of range in the view.

Done

**Pull request merged**